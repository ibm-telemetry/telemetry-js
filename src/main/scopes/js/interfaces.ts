/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
