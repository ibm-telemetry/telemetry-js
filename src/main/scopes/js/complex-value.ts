/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
