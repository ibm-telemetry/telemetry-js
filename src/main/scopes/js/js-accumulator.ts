/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
