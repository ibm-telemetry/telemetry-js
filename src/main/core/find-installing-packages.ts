/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { NoNodeModulesFoundError } from '../exceptions/no-node-modules-found-error.js'
import { findInstallersFromTree } from '../scopes/npm/find-installers-from-tree.js'
import { hasNodeModulesFolder } from '../scopes/npm/has-node-modules-folder.js'
import { InstallingPackage, PackageData } from '../scopes/npm/interfaces.js'
import { DirectoryEnumerator } from './directory-enumerator.js'
import { Logger } from './log/logger.js'
import { runCommand } from './run-command.js'

/**
 * Finds all packages within the project that installed the specified package at the specified
 * version. This is done by starting at the current directory and traversing up the directory
 * structure until an `npm ls` command on one of those directories returns a non-empty list of
 * installers.
 *
 * If no installers were found after the root-most project directory was searched, an empty array
 * is returned.
 *
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param packageData - The name  and exact version of the package to search for.
 * @param logger - Logger instance.
 * @returns A possibly empty array of installing packages.
 */
export async function findInstallingPackages(
  cwd: string,
  root: string,
  packageData: PackageData,
  logger: Logger
): Promise<InstallingPackage[]> {
  logger.traceEnter('', 'findInstallingPackages', [cwd, root, packageData])
  const dirs = await new DirectoryEnumerator(logger).find(cwd, root, hasNodeModulesFolder)
  const topMostDir = dirs.pop()

  if (topMostDir === undefined) {
    throw new NoNodeModulesFoundError(cwd, root)
  }

  // Allow this command to try and obtain results even if it exited with a total or partial error
  const result = await runCommand('npm ls --all --json', logger, { cwd: topMostDir }, false)

  const dependencyTree = JSON.parse(result.stdout)

  const installers = findInstallersFromTree(dependencyTree, packageData, logger)

  logger.traceExit('', 'findInstallingPackages', installers)
  return installers
}
