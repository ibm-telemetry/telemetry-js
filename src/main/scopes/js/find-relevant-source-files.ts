/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
