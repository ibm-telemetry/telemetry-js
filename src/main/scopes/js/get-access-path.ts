/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
