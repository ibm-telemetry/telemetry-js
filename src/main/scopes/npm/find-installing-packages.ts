/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { exec } from '../../core/exec.js'
import { getProjectRoot } from '../../core/get-project-root.js'
import { findInstallersFromTree } from './find-installers-from-tree.js'
import { findScannableDirectories } from './find-scannable-directories.js'
import { type InstallingPackage } from './interfaces.js'

/**
 * Finds all packages within the project that installed the specified package at the specified
 * version. This is done by starting at the current directory and traversing up the directory
 * structure until an `npm ls` command on one of those directories returns a non-empty list of
 * installers.
 *
 * If no installers were found after the root-post project directory was searched, an empty array is
 * returned.
 *
 * @param packageName - The name of the package to search for.
 * @param packageVersion - The exact version of the package to search for.
 * @returns A possibly empty array of installing packages.
 */
export async function findInstallingPackages(
  packageName: string,
  packageVersion: string
): Promise<InstallingPackage[]> {
  const cwd = process.cwd()
  const root = await getProjectRoot()
  const dirs = await findScannableDirectories(cwd, root)

  let installers: InstallingPackage[] = []

  for (const d of dirs) {
    const dependencyTree = JSON.parse(await exec('npm ls --all --json', { cwd: d }))

    installers = findInstallersFromTree(dependencyTree, packageName, packageVersion)

    if (installers.length > 0) {
      break
    }
  }

  return installers
}
