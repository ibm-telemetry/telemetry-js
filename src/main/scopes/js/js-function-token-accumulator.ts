/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
