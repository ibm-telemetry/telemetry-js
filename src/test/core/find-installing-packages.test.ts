/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path'
import { describe, expect, it } from 'vitest'

import { findInstallingPackages } from '../../main/core/find-installing-packages.js'
import { NoPackageJsonFoundError } from '../../main/exceptions/no-package-json-found-error.js'
import { Fixture } from '../__utils/fixture.js'
import { initLogger } from '../__utils/init-logger.js'

describe('findInstallingPackages', () => {
  const logger = initLogger()
  it('throws an error if no package.json files were found', async () => {
    const fixture = new Fixture('projects/no-package-json-files/foo/bar')

    await expect(
      async () =>
        await findInstallingPackages(
          fixture.path,
          path.join(fixture.path, '..', '..'),
          logger,
          'not-here',
          '0.1.0'
        )
    ).rejects.toThrow(NoPackageJsonFoundError)
  })

  it('correctly finds installing package data', async () => {
    const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
    const pkgs = await findInstallingPackages(
      fixture.path,
      path.join(fixture.path, '..', '..'),
      logger,
      'instrumented',
      '0.1.0'
    )

    expect(pkgs).toMatchSnapshot()
  })

  it('finds no results for an unknown package', async () => {
    const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
    const pkgs = await findInstallingPackages(
      fixture.path,
      path.join(fixture.path, '..', '..'),
      logger,
      'not-here',
      '0.1.0'
    )

    expect(pkgs).toMatchSnapshot()
  })

  it('finds no results for a known package at an unknown version', async () => {
    const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
    const pkgs = await findInstallingPackages(
      fixture.path,
      path.join(fixture.path, '..', '..'),
      logger,
      'instrumented',
      '0.3.0'
    )

    expect(pkgs).toMatchSnapshot()
  })
})
