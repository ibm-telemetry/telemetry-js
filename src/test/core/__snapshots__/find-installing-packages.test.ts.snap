// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findInstallingPackages > correctly finds installing package data 1`] = `
[
  {
    "dependencies": [
      {
        "name": "foo",
        "version": "1.0.0",
      },
      {
        "name": "instrumented",
        "version": "0.1.0",
      },
    ],
    "name": "basic-project",
    "version": "1.0.0",
  },
]
`;

exports[`findInstallingPackages > finds no results for a known package at an unknown version 1`] = `[]`;

exports[`findInstallingPackages > finds no results for an unknown package 1`] = `[]`;
